Add the initialisation script and provide initial description in the README

hadrianvalentine opened the pull request:
# Description
The idea behind this script is that it will run in a golang container as a sidecar in a vault pod.

It will:
1. Make an `init` request to Vault.
2. Receive the root token and encrypt it using AWS KMS.
3. Upload the encrypted token to S3.

Note: We receive other items like recovery keys in the init response. These can also be added if necessary as there already recorded in the `InitResponse` as you'll see in the code.

## Inspiration and examples drawn from:
https://gist.github.com/fuzzyami/f3a7231037166117a6fef9607960aee7
https://gist.github.com/andreas/eea322d73ca21e66b55b
https://docs.aws.amazon.com/sdk-for-go/api/service/s3/
https://godoc.org/github.com/aws/aws-sdk-go/service/s3/s3manager

hadrianvalentine committed:
Add the initialisation script and provide initial description in the README

hadrianvalentine committed:
Add MIT license

hadrianvalentine committed:
Move code into Cobra CLI structure.

hadrianvalentine committed:
Add Golang linter

hadrianvalentine committed:
Add Dockerfile and dockerignore

hadrianvalentine committed:
Add dependencies and test to the Makefile

hadrianvalentine committed:
Add dependency file

hadrianvalentine committed:
Add Travis CI file

hadrianvalentine committed:
Rename to docker-vault-init for consistency

hadrianvalentine committed:
Install docker-vault-init on container

hadrianvalentine committed:
Remove viper

hadrianvalentine committed:
fix InitResponse struct

hadrianvalentine committed:
Change commenting to address linter errors

hadrianvalentine committed:
No underscores in varable names

hadrianvalentine committed:
Install dep

hadrianvalentine committed:
Address linting warning

hadrianvalentine committed:
Add linting stage to CI

hadrianvalentine committed:
Install tools before linting

itskingori commented on README.md:
```diff
@@ -1 +1,2 @@
-# docker-vault-init
\ No newline at end of file
+# Docker Vault Init
```
I think the `README` needs some instructions just so that someone who's new to the project can get up to speed without having to figure out a lot by looking at the code. Something like (to be modified appropriately) ...

```markdown
## Development

Below instructions are only necessary if you intend to work on the source code.
For normal usage the above installation instruction should do.

### Requirements

1. Ensure that you have a [properly
   configured](https://golang.org/doc/code.html#Workspaces) Go workspace.
1. For dependency management, you'll need `dep` which can be installed with
   `brew install dep`.

### Building

1. Fetch the code with `go get -v github.com/Intellection/docker-vault-init`.
1. Install the Go development tools via `make tools`.
1. Install `dep` via `brew install dep`.
1. Install application dependencies via `make dependencies` (they'll be placed
   in `./vendor`). Requires [golang/dep][dep] package manager.
1. Build and install the binary with `make build tag=x.y.z`.
1. Run the command e.g. `./bin/vault-init help` as a basic test.

### Testing

1. Install the Go testing tools via `make tools`.
2. Run linter using `make lint` and test using `make test`.
```

hadrianvalentine replied:
Updated with suggested dev info here https://github.com/Intellection/vault-init/pull/1/commits/58e355dd2cf8c3be23b72961de69c3a80bdd57f1

itskingori commented on cmd/initialise.go:
```diff
@@ -0,0 +1,189 @@
+var initialiseCmd = &cobra.Command{
+	Use:   "initialise",
+	Short: "A brief description of your command",
+	Long: `A longer description that spans multiple lines and likely contains examples
```
This needs to be updated.

hadrianvalentine replied:
Updated in https://github.com/Intellection/vault-init/pull/1/commits/282eff9e6cb1740337fa647d1fdb9102b7369054

itskingori commented on cmd/initialise.go:
```diff
@@ -0,0 +1,189 @@
+// initialiseCmd represents the initialise command
+var initialiseCmd = &cobra.Command{
+	Use:   "initialise",
+	Short: "A brief description of your command",
```
This needs to be updated.

hadrianvalentine replied:
Updated in https://github.com/Intellection/vault-init/pull/1/commits/282eff9e6cb1740337fa647d1fdb9102b7369054

itskingori commented on cmd/initialise.go:
```diff
@@ -0,0 +1,189 @@
+		fmt.Println("Initialisation complete")
+
+		// AWS setup
+		region := "us-east-1"
```
This could be parametrised into a flag whose value defaults to `us-east-1`. You don't have to do this now. But file it as an issue that can be closed when done.

hadrianvalentine replied:
Can do. I was wondering about whether to put [this config value](https://github.com/Intellection/vault-init/pull/1/files#diff-6bf4c7bc7dd75a561cdd357f83883607R124) in a flag as well instead of getting from the env vars. Thoughts?

itskingori replied:
Ok with envars as well.

hadrianvalentine replied:
Alright. Put this in an env var as well
https://github.com/Intellection/vault-init/pull/1/commits/c84af2aa15a2064942d569c4164832c2a78352ec

itskingori commented on cmd/initialise.go:
```diff
@@ -0,0 +1,189 @@
+func initVault() string {
+	vaultAddr := os.Getenv("VAULT_ADDR")
+	if vaultAddr == "" {
+		vaultAddr = "https://127.0.0.1:8200"
```
For local, the default should be `http` since we don't have certificates locally, right?

```suggestion
		vaultAddr = "http://127.0.0.1:8200"
```

itskingori commented on cmd/initialise.go:
```diff
@@ -0,0 +1,189 @@
+	httpClient = http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
```
Should we do this? Maybe this was done to handle the possible error cause by having `https://127.0.0.1:8200` above. 🤔 

```suggestion
				InsecureSkipVerify: false,
```

hadrianvalentine replied:
Not sure to be honest. Let's give it a go.

hadrianvalentine replied:
Not sure to be honest. Let's give it a go.

itskingori commented on cmd/initialise.go:
```diff
@@ -0,0 +1,189 @@
+	rootCmd.AddCommand(initialiseCmd)
+}
+
+func checkError(e error) {
```
LOL, welcome to Go! 😅

hadrianvalentine replied:
😅

itskingori reviewed (commented):
Some initial comments 💬

itskingori commented:
FYI, renamed repo from `Intellection/docker-vault-init` to `vault-init`, so you'll need to change references accordingly (they are about 14 occurrences).

hadrianvalentine committed:
Use HTTP for local

Co-Authored-By: King'ori Maina <j@kingori.co>

hadrianvalentine committed:
Add short and long introductions for initialise command.

hadrianvalentine committed:
Update references to the package and repo name to reflect the change in the repo name.

hadrianvalentine committed:
Pull AWS region from environment variable.

hadrianvalentine commented:
Just realised the name change means it's not being followed by Travis anymore. Fixing.

hadrianvalentine committed:
Disable skipping verification for TLS

Co-Authored-By: King'ori Maina <j@kingori.co>

hadrianvalentine committed:
Update README with development instructions.

itskingori commented:
@hadrianvalentine I think we need a `.gitignore` that excludes `vendor/` and `bin/`. To avoid this ...

<img width="1680" alt="Screenshot 2019-10-09 at 08 12 20" src="https://user-images.githubusercontent.com/703786/66453452-0a7fe700-ea6d-11e9-9677-269bba98b891.png">

Also, I think we should check in `Gopkg.lock` as well (seen in screenshot above).

itskingori commented on README.md:
```diff
@@ -1 +1,28 @@
-# docker-vault-init
\ No newline at end of file
+# Vault Init
```
I couldn't find the project on Travis but I'm guessing that's because we need the `.travis.yml` in `master for the integration to kick in. Since the badges are predictably constructed, I guess we can get this in now.

```suggestion
# Vault Init

[![Build Status](https://travis-ci.org/Intellection/vault-init.svg?branch=master)](https://travis-ci.org/Intellection/vault-init)

```

itskingori replied:
Ah, turns out I was looking in travis-ci.com instead of travis-ci.org 🤦‍♂ ... updated suggestion.

hadrianvalentine replied:
I ended up accidentally creating an account on `travis-ci.com` 🤦‍♂ 
It is very misleading 😅

hadrianvalentine replied:
So...
https://devops.stackexchange.com/questions/1201/whats-the-difference-between-travis-ci-org-and-travis-ci-com?answertab=votes#tab-top

itskingori replied:
Looks like you should be using travis-ci.com. 😅

hadrianvalentine replied:
![3cs7i9](https://user-images.githubusercontent.com/17402708/66475854-79bd0180-ea94-11e9-8339-acdc97c80940.jpg)

itskingori reviewed (commented):
Left a comment related to Travis inline.

That said, you can also think about how to use Travis for deployment. This is how I've done it in my own project [itskingori/sanaa@0.10.0/.travis.yml](https://github.com/itskingori/sanaa/blob/0.10.0/.travis.yml) so that on tag, it builds the binary, creates a release on GitHub and then triggers a build on Docker Hub. You don't have to do this specific part on this PR.

itskingori commented on Makefile:
```diff
@@ -0,0 +1,26 @@
+	go get -u github.com/alecthomas/gometalinter
+
+	# install gox
+	go get -v github.com/mitchellh/gox
```
Just a note that you have this in but we don't actually use it. But you can leave it in since you'll probably integrate release on tag (which isn't in this PR yet, see https://github.com/Intellection/vault-init/pull/1#pullrequestreview-299177970).

itskingori commented on Makefile:
```diff
@@ -0,0 +1,26 @@
+	@mkdir -p bin/
+	go build -o ./bin/vault-init
+	@echo
+	@./bin/vault-init version
```
I think we should have a space below here as above ...

```suggestion

	@./bin/vault-init version
```

itskingori commented on Makefile:
```diff
@@ -0,0 +1,26 @@
+	@echo
+	@./bin/vault-init version
+dependencies:
+	dep ensure
```
I think we should have a space below here as above ...

```suggestion

	dep ensure
```

itskingori reviewed (commented):
Some spacing comments. You can accept as batch so that they are combined into one commit.

itskingori commented on README.md:
```diff
@@ -1 +1,28 @@
+1. Install `dep` via `curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh`. `brew install dep` can also be used if using macOS.
+1. Install application dependencies via `make dependencies` (they'll be placed
+   in `./vendor`). Requires [golang/dep][dep] package manager.
+1. Build and install the binary with `make build tag=x.y.z`.
```
This step fails as shown below ...

```console
$ make build tag=0.0.0
go build -o ./bin/vault-init

Error: unknown command "version" for "vault-init"
Run 'vault-init --help' for usage.
unknown command "version" for "vault-init"
make: *** [build] Error 1
```

itskingori commented on cmd/initialise.go:
```diff
@@ -0,0 +1,190 @@
+// initialiseCmd represents the initialise command
+var initialiseCmd = &cobra.Command{
+	Use:   "initialise",
+	Short: "Intitial a specified instance of Vault",
```
```suggestion
	Short: "Intitialise a specified instance of Vault",
```

itskingori commented on cmd/initialise.go:
```diff
@@ -0,0 +1,190 @@
+var initialiseCmd = &cobra.Command{
+	Use:   "initialise",
+	Short: "Intitial a specified instance of Vault",
+	Long: `This command makes a requets to instance of Vault at the specified Vault address
```
```suggestion
	Long: `This command makes a request to an instance of Vault at the specified Vault address
```

itskingori commented on cmd/initialise.go:
```diff
@@ -0,0 +1,190 @@
+been setup to occur during Vault initialisation.
+
+Once it has received the token in the response from Vault, it will encrypt and
+store this token on S3 when it can be used for authentication by entities that
```
```suggestion
store this token on S3 from where it can be used for authentication by entities that
```

itskingori reviewed (commented):
Typos in the description

itskingori commented on cmd/initialise.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+func fullKeyID(accountID string, keyID string, region string) (string) {
+	baseString := fmt.Sprintf("arn:aws:kms:region:%s:key/%s", accountID, keyID)
```
Looking at the function signature, do we need to replace `region` here?

hadrianvalentine replied:
Yes, you're right.

itskingori commented on cmd/initialise.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+func init() {
+	rootCmd.AddCommand(initialiseCmd)
```
Do we really need a `initialise` command?

```console
$ bin/vault-init -h
A longer description that spans multiple lines and likely contains
examples and usage of using your application. For example:

Cobra is a CLI library for Go that empowers applications.
This application is a tool to generate the needed files
to quickly create a Cobra application.

Usage:
  vault-init [command]

Available Commands:
  help        Help about any command
  initialise  Intitial a specified instance of Vault

Flags:
      --config string   config file (default is $HOME/.vault-init.yaml)
  -h, --help            help for vault-init
  -t, --toggle          Help message for toggle

Use "vault-init [command] --help" for more information about a command.
```

As is, we would need to do `vault-init initialise`, which seems unnecessary (considering the purpose of this tool is only to initialise vault), we could just do, `vault init`. Also means that a lot of what's here can go to `cmd/root.go`.

hadrianvalentine replied:
We probably don't. we could just move it to root, particularly since we don't need to pass in arguments.

hadrianvalentine replied:
Moved to root in https://github.com/Intellection/vault-init/pull/1/commits/f0800f6d69a05a4c616c8b4e52813195066b5328

itskingori commented on cmd/root.go:
```diff
@@ -0,0 +1,102 @@
+  // Cobra supports persistent flags, which, if defined here,
+  // will be global for your application.
+
+  rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default is $HOME/.vault-init.yaml)")
```
I see we have a global flag here ... do we use it?

```
Global Flags:
      --config string   config file (default is $HOME/.vault-init.yaml)
```

hadrianvalentine replied:
Na, don't need it.

hadrianvalentine replied:
Removed in https://github.com/Intellection/vault-init/pull/1/commits/62f4daedbe8be2955df96e709f9fdafe5a72b8cc

itskingori commented on cmd/root.go:
```diff
@@ -0,0 +1,102 @@
+
+  // Cobra also supports local flags, which will only run
+  // when this action is called directly.
+  rootCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
```
This looks like boilerplate code. Should be ok to remove.

hadrianvalentine replied:
Killing

hadrianvalentine replied:
Removed in https://github.com/Intellection/vault-init/pull/1/commits/62f4daedbe8be2955df96e709f9fdafe5a72b8cc

itskingori reviewed (commented):
Some more comments around cobra configuration.

itskingori commented:
Could you add a usage section with a basic example? For your reference see the "Installation", "Configuration" and "Usage" sections in [itskingori/sanaa@0.10.0/docs/index.md](https://github.com/itskingori/sanaa/blob/0.10.0/docs/index.md).

itskingori commented:
Just remembered another thing, could you add a `CODEOWNERS` file?

hadrianvalentine committed:
Ignore vender and bin directories in git

hadrianvalentine committed:
Add dependency lock file.

hadrianvalentine committed:
Command descriptions typo

Co-Authored-By: King'ori Maina <j@kingori.co>

hadrianvalentine committed:
Update long description

Co-Authored-By: King'ori Maina <j@kingori.co>

hadrianvalentine committed:
Update long description

Co-Authored-By: King'ori Maina <j@kingori.co>

hadrianvalentine committed:
Update Makefile

Co-Authored-By: King'ori Maina <j@kingori.co>

hadrianvalentine committed:
Add CI badge

Co-Authored-By: King'ori Maina <j@kingori.co>

hadrianvalentine committed:
Update Makefile

Co-Authored-By: King'ori Maina <j@kingori.co>

hadrianvalentine committed:
Add region to formatted string

hadrianvalentine committed:
Merge branch 'add_init_script' of github.com:Intellection/vault-init into add_init_script

hadrianvalentine committed:
Remove Cobra boilerplate code.

hadrianvalentine committed:
Add CODEOWNERS file

hadrianvalentine commented:
>Just remembered another thing, could you add a CODEOWNERS file?

Done in https://github.com/Intellection/vault-init/pull/1/commits/462cd5d21aa4144c81ffd9db169dd394fe3e8b27

hadrianvalentine commented:
> Could you add a usage section with a basic example? For your reference see the "Installation", "Configuration" and "Usage" sections in itskingori/sanaa@0.10.0/docs/index.md.

Will add after addressing [this comment](https://github.com/Intellection/vault-init/pull/1#discussion_r332846381) and moving the logic to root.

hadrianvalentine committed:
Move initialise logic to root.

itskingori commented on Makefile:
```diff
@@ -0,0 +1,28 @@
+	go build -o ./bin/vault-init
+	@echo
+
+	@./bin/vault-init version
```
Continuing from https://github.com/Intellection/vault-init/pull/1/files#r332842914, this will fail because there's no `version` command so it's taken as an argument to the root command which tries to talk to Vault. In my case I don't have it running so it fails but at least it highlights the issue. See:

```console
$ make build tag=0.0.0
go build -o ./bin/vault-init

Initialising Vault...
panic: Put http://127.0.0.1:8200/v1/sys/init: dial tcp 127.0.0.1:8200: connect: connection refused

goroutine 1 [running]:
github.com/Intellection/vault-init/cmd.checkError(...)
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/cmd/root.go:162
github.com/Intellection/vault-init/cmd.initVault(0x0, 0x0)
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/cmd/root.go:195 +0x673
github.com/Intellection/vault-init/cmd.glob..func1(0x1e7c320, 0xc000069c40, 0x1, 0x1)
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/cmd/root.go:79 +0x9b
github.com/Intellection/vault-init/vendor/github.com/spf13/cobra.(*Command).execute(0x1e7c320, 0xc00000e070, 0x1, 0x1, 0x1e7c320, 0xc00000e070)
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/vendor/github.com/spf13/cobra/command.go:830 +0x2aa
github.com/Intellection/vault-init/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x1e7c320, 0x103cc2a, 0x1e3e0c0, 0xc000000180)
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/vendor/github.com/spf13/cobra/command.go:914 +0x2fb
github.com/Intellection/vault-init/vendor/github.com/spf13/cobra.(*Command).Execute(...)
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/vendor/github.com/spf13/cobra/command.go:864
github.com/Intellection/vault-init/cmd.Execute()
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/cmd/root.go:123 +0x31
main.main()
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/main.go:26 +0x20
make: *** [build] Error 2
```

Recommended solution, just remove it and keep just the building. We usually do this just to show that the thing works.

hadrianvalentine replied:
Removed in https://github.com/Intellection/vault-init/pull/1/commits/259f0f00ac7d7a7986c4ac8126b21023c4daaff3

```
$ make build tag=0.0.0
go build -o ./bin/vault-init


```

itskingori commented on Makefile:
```diff
@@ -0,0 +1,28 @@
+
+	@./bin/vault-init version
+dependencies:
+
```
I might have fudged up on the suggestion here (GitHub UI is a bit weird on multi-line suggestions. Could you review the spacing in the entire file.

hadrianvalentine replied:
Cool. So we'll remove the extra spaces that were added after the commands?

itskingori replied:
Yes ... like `tools` and `lint` 👇 

```makefile
SHELL := /bin/bash

.PHONY: tools lint build dependencies test

tools:
	# install gometalinter
	go get -u github.com/alecthomas/gometalinter

	# install gox
	go get -v github.com/mitchellh/gox

	# install all known linters:
	gometalinter --install

lint:
	gometalinter --config="linters.json" ./...

build:
	@mkdir -p bin/
	go build -o ./bin/vault-init
	@echo

dependencies:
	dep ensure

test:
	go test -v ./...
```

itskingori replied:
This hasn't been addressed (also edited previous comment who's example was wrong ... forgot to edit after copy).

itskingori commented on cmd/root.go:
```diff
@@ -0,0 +1,234 @@
+	// If a config file is found, read it in.
+	if err := viper.ReadInConfig(); err == nil {
+		fmt.Println("Using config file:", viper.ConfigFileUsed())
+	}
```
I don't think we need anything here since we're not reading configuration from anywhere. Should have been removed as part of addressing https://github.com/Intellection/vault-init/pull/1#discussion_r332846779.

itskingori commented on cmd/root.go:
```diff
@@ -0,0 +1,234 @@
+
+
+var httpClient    http.Client
+
```
Extra space ...

```suggestion
```

itskingori commented on cmd/root.go:
```diff
@@ -0,0 +1,234 @@
+	cobra.OnInitialize(initConfig)
+}
+
+
```
Extra space ...

```suggestion
```

itskingori commented on cmd/root.go:
```diff
@@ -0,0 +1,234 @@
+		log.Println(errUn)
+		panic(errUn)
+	}
+
```
Extra space ...

```suggestion
```

itskingori commented on cmd/root.go:
```diff
@@ -0,0 +1,234 @@
+	baseString := fmt.Sprintf("arn:aws:kms:%s:%s:key/%s", region, accountID, keyID)
+	return baseString
+}
+
```
Extra space ...

```suggestion
```

itskingori reviewed (commented):
We're getting there. Some more comments (see inline). That said, are you still planning on doing the release stuff in this PR?

I've left a bunch of extra space suggestions that you can accept as a batch so that they are in one commit ... 👇 

![add-suggestion-to-batch](https://user-images.githubusercontent.com/703786/66625481-07b3fc00-ebfd-11e9-95b8-520f7ec56cf7.png)

hadrianvalentine committed:
Apply spacing suggestions from code review

Co-Authored-By: King'ori Maina <j@kingori.co>

hadrianvalentine committed:
Adjust spacing as per review feedback

hadrianvalentine committed:
Remove Viper

hadrianvalentine committed:
Update dependencies

hadrianvalentine committed:
Add script for cross compilation.

hadrianvalentine committed:
Add remote Docker Hub trigger

hadrianvalentine commented on scripts/trigger.sh:
```diff
@@ -0,0 +1,8 @@
+set -eu
+
+# trigger docker hub
+curl -X "POST" "https://cloud.docker.com/api/build/v1/source/${DOCKER_HUB_TRIGGER_SOURCE}/trigger/${DOCKER_HUB_TRIGGER_ID}/call/" \
```
@itskingori I've added a trigger for docker hub. Seems different from the one you used before.

<img width="1089" alt="Screenshot 2019-10-11 at 16 32 40" src="https://user-images.githubusercontent.com/17402708/66660282-3f469680-ec45-11e9-939c-17164354d908.png">

itskingori replied:
Should be fine. Each repo should have it's own unique hook. That said, can't see anything ...

<img width="1136" alt="Screenshot 2019-10-14 at 07 51 55" src="https://user-images.githubusercontent.com/703786/66730012-8ad08e80-ee57-11e9-8e52-cc4bb39f6626.png">

itskingori replied:
Related, https://github.com/Intellection/vault-init/pull/1#issuecomment-541502504.

itskingori replied:
@hadrianvalentine I think you might also have to specify the build rules, something like this ...

<img width="1221" alt="Screenshot 2019-11-06 at 10 27 39" src="https://user-images.githubusercontent.com/703786/68277487-5a5bc900-0080-11ea-854e-e04c329317fd.png">

If you go check now you'll see there are none set and they give you some examples:

<img width="1221" alt="Screenshot 2019-11-06 at 10 30 58" src="https://user-images.githubusercontent.com/703786/68277678-bcb4c980-0080-11ea-80cc-6886c0455f09.png">

It tells Docker Hub what to use for the tag from the version passed in (which must match the regex). For example, in the rules above, it'll create two docker tags, `latest` and `<version>` for each trigger.

itskingori commented on README.md:
```diff
@@ -1 +1,31 @@
+1. Install application dependencies via `make dependencies` (they'll be placed
+   in `./vendor`). Requires [golang/dep][dep] package manager.
+1. Build and install the binary with `make build tag=x.y.z`.
+1. Run the command e.g. `./bin/vault-init help` as a basic test.
```
This should be removed since there's no `help` command ...

```console
$ ./bin/vault-init help
Initialising Vault...
panic: Put http://127.0.0.1:8200/v1/sys/init: dial tcp 127.0.0.1:8200: connect: connection refused

goroutine 1 [running]:
github.com/Intellection/vault-init/cmd.checkError(...)
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/cmd/root.go:131
github.com/Intellection/vault-init/cmd.initVault(0x0, 0x0)
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/cmd/root.go:164 +0x673
github.com/Intellection/vault-init/cmd.glob..func1(0x1d02f00, 0xc0000bb780, 0x1, 0x1)
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/cmd/root.go:75 +0x9b
github.com/Intellection/vault-init/vendor/github.com/spf13/cobra.(*Command).execute(0x1d02f00, 0xc0000cc010, 0x1, 0x1, 0x1d02f00, 0xc0000cc010)
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/vendor/github.com/spf13/cobra/command.go:830 +0x2aa
github.com/Intellection/vault-init/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x1d02f00, 0x103cb9a, 0x1cc8760, 0xc000000180)
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/vendor/github.com/spf13/cobra/command.go:914 +0x2fb
github.com/Intellection/vault-init/vendor/github.com/spf13/cobra.(*Command).Execute(...)
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/vendor/github.com/spf13/cobra/command.go:864
github.com/Intellection/vault-init/cmd.Execute()
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/cmd/root.go:119 +0x31
main.main()
        /Users/Kingori/Workspace/src/github.com/Intellection/vault-init/main.go:26 +0x20
```

itskingori commented on cmd/root.go:
```diff
@@ -0,0 +1,201 @@
+	"os"
+)
+
+
```
Missed this extra space ...

```suggestion
```

itskingori commented on cmd/root.go:
```diff
@@ -0,0 +1,201 @@
+}
+
+func init() {
+
```
I suspect cobra needs this. Just a suggestion ... could we change this to `func init() {}`?

itskingori commented:
Seems like you didn't set the scripts to be executable so they won't be able to run ...

```console
$ ls -l scripts/*.sh
-rw-r--r--  1 Kingori  staff  1695 14 Oct 07:33 scripts/compile.sh
-rw-r--r--  1 Kingori  staff   185 14 Oct 07:33 scripts/trigger.sh
```

itskingori commented:
TravisCI configuration for deployments seems to be missing from `.travis.yml`.

hadrianvalentine committed:
Remove whitespace.

Co-Authored-By: King'ori Maina <j@kingori.co>

hadrianvalentine committed:
Remove failing command from README.

hadrianvalentine committed:
Style

hadrianvalentine committed:
Make scripts executable

itskingori commented:
> TravisCI configuration for deployments seems to be missing from `.travis.yml`.

This hasn't been addressed.

itskingori reviewed (dismissed):
I'm gonna approve. If there are any issues you can fix them in another PR. That said, see:

* Some missing config on the docker-hub side - https://github.com/Intellection/vault-init/pull/1#discussion_r342949378
* Some missing config from `.travis.yml` - https://github.com/Intellection/vault-init/pull/1#issuecomment-550185152

hadrianvalentine committed:
Add deploy to travis file

hadrianvalentine commented:
> Some missing config on the docker-hub side

Added a rule for tags.

> Some missing config from .travis.yml - #1 (comment)

Addressed here https://github.com/Intellection/vault-init/pull/1/commits/6455e9b9d945e1dbbc45032afbcdce83fd4569da

itskingori commented on .travis.yml:
```diff
@@ -0,0 +1,32 @@
+language: go
```
I know this is not mandatory but could we have the `---` that was there before? It's common/standard in all of our YAML files. For [an explanation why, see here](https://stackoverflow.com/a/50788318). If you feel strongly not to add it, then it's cool, you can leave it out.

hadrianvalentine replied:
Done https://github.com/Intellection/vault-init/pull/1/commits/55cbd25c7bc0d4c0d2c6c5eb8f31714b5da96f2b

itskingori commented on .travis.yml:
```diff
@@ -0,0 +1,32 @@
+language: go
+go:
+- 1.13.x
```
Could we indent arrays by two spaces (off the key under which they are). Understandably, this ... to my surprise, is still valid YAML. 🙄

This applies in other instances below e.g. `before_install` etc.

hadrianvalentine replied:
Done https://github.com/Intellection/vault-init/pull/1/commits/2c564f7c5b31ca61115fe52902930c2ad978c958

itskingori replied:
So far so good ... but I'm feeling guilty because I keep commenting of stuff that fundamentally doesn't matter, but there's still some more stuff you didn't indent. See `include:` and `before_deploy:`.

hadrianvalentine replied:
Ah, missed it. Thank you. Put all the indents in one commit https://github.com/Intellection/vault-init/pull/1/commits/5228613499a350a3eb72c5563225abcf80fc2c2b

itskingori commented on .travis.yml:
```diff
@@ -0,0 +1,32 @@
+    prerelease: false
+    on:
+      condition: ${TRAVIS_TAG} =~ ^[0-9]+\.[0-9]+\.[0-9]+$
+      repo: itskingori/sanaa
```
Copy-pasta 😅 ...

```suggestion
      repo: Intellection/vault-init
```

hadrianvalentine replied:
Haha 😅

itskingori reviewed (commented):
Some comments ... 💬

hadrianvalentine committed:
Use correct repo name.

Co-Authored-By: King'ori Maina <j@kingori.co>

hadrianvalentine committed:
Use dashes in YAML to indicate the start of the document

hadrianvalentine committed:
Indent arrays.

itskingori reviewed (approved):
👌
